compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,203 @@
+        // we only allow this optimization on locals with no projections.
+        let Some(switched) = switched.as_local() else {
+            continue
+        };
+
+        // We also do not allow locals that are borrowed as they could be
+        // indirectly mutated before the switch.
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some((discr, index)) = data.statements.iter().enumerate().rev().find_map(
+            |(index, statement)| {
+                match statement.kind {
+                    StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                        if place.as_local() == Some(switched) =>
+                    {
+                        Some((discr, index))
+                    }
+                    _ => None,
+                }
+            }
+        ) else {
+            continue
+        };
+
+        let Some(discr) = discr.as_local() else {
+            continue
+        };
+
+        if borrowed_locals.contains(discr) {
+            continue;
+        }
+
+        // We need to make sure that `switched` is not overwritten before the switch,
+        // and also that `discr` is not overwritten before the `Rvalue::Discriminant`.
+        if is_local_mutated(body, block, discr, index, 0..index)
+            || is_local_mutated(body, block, switched, index, index..data.statements.len())
+        {
+            continue;
+        }
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
+                    StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place.as_local() == Some(discr) => {
+                        debug!("{pred:?}: {place:?} = {def_id:?}::{variant:?}; goto -> {block:?};");
+                        let discr_ty = tcx.type_of(def_id).subst(tcx, substs);
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            let target = targets.target_for_value(discr.val);
+                            new_targets.entry(block).or_insert_with(Vec::new).push((pred, target));
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    if new_targets.is_empty() {
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    for (block, new_targets) in new_targets {
+        let statements: Vec<_> = basic_blocks[block].statements.iter().cloned().collect();
+        for (pred, target) in new_targets {
+            let pred = &mut basic_blocks[pred];
```

Comment:
Please document why there is only a single `pred -> block` edge in the whole CFG, and that no earlier threading has overwritten it.

Reply:
It may also be useful to document why `pred != block` and assert it.